p2p/peer.go

```diff
@@ -18,6 +18,7 @@ import (
 // Peer is an interface representing a peer connected on a reactor.
 type Peer interface {
 	cmn.Service
+	QuitChan() <-chan struct{}
```

Comment:
Why do we need this new QuitChan? isn't `Quit` already exposed through the Service ?

Reply:
```
type Service interface {
	// Start the service.
	// If it's already started or stopped, will return an error.
	// If OnStart() returns an error, it's returned by Start()
	Start() error
	OnStart() error

	// Stop the service.
	// If it's already stopped, will return an error.
	// OnStop must never error.
	Stop() error
	OnStop()

	// Reset the service.
	// Panics by default - must be overwritten to enable reset.
	Reset() error
	OnReset() error

	// Return true if the service is running
	IsRunning() bool

	// String representation of the service
	String() string

	SetLogger(log.Logger)
}
```

Reply:
we need it in order to exit the broadcast routine when peer stops:

```go
case <-memR.Mempool.TxsWaitChan(): // Wait until a tx is available
				if next = memR.Mempool.TxsFront(); next == nil {
					continue
				}
			case <-peer.QuitChan():
				return
			case <-memR.Quit:
				return
			}
```

Reply:
I had posted this on the internal engineering channel, but here it is for posterity
`QuitChan() <- chan reason` exposes that the underlying implementation:
a) leaks to the rest of the world that the underlying implementation always uses a channel
b) gives the caller access to abuse the channel e.g by closing it and then causing a runtime panic if someone else tries to send on it
c) Limits what the Quit method will always do. What if we just wanted to log an action, or send an HTTP POST request on Quit?
d) Will expose to the entire world that perhaps they need to run Quit in a goroutine just in case the QuitChan is not actually buffered :disappointed:
Perhaps we can let the underlying code use a channel and then wrap that functionality up in a nice method that
a) triggers channel closing and not worrying about buffering
b) allows the underlying implementation to do whatever it wants e.g
```go
func (p *peerUno) Quit() {
    peerUno.closeOnce.Do(func() {
        close(peerUno.quitChan)
        peerUno.log.Logf("Quiting now!")
        peerUno.Flush()
    }
}
```

Reply:
oh so let's make peer wrap BaseService instead of Service ?

Reply:
Peer interface wrapping BaseService struct?

Reply:
Yeh why not ? We do that in lots of places.

Reply:
Jae suggested we include `Quit() <-chan struct{}` method into `Service` interface, so maybe we should just do that

Reply:
@odeke-em the problem is that we need to be able to break out of blocking on a channel using a `select` block and a `<- quitChan`. So for that reason we expose that it's a channel.

Ok @melekes I think this is fine then. We'll use the `service.Quit()` when we finish the Service refactor ie https://github.com/tendermint/tmlibs/pull/114

Reply:
lets s/QuitChan/s/Quit/g for consistency.  Quit is always a chan.